lib/ansible/module_utils/basic.py

```diff
@@ -427,6 +432,88 @@ def remove_values(value, no_log_strings):
     return new_value
 
 
+def sanitize_keys(obj, no_log_strings, ignore_keys=None):
+    """ Sanitize the keys in a container object by removing no_log values from key names.
+
+    This is a companion function to the `remove_values()` function. Similar to that function,
+    we make use of deferred_removals to avoid hitting maximum recursion depth in cases of
+    large data structures.
+
+    :param obj: The container object to sanitize. Non-container objects are returned unmodified.
+    :param no_log_strings: A set of string values we do not want logged.
+    :param ignore_keys: A set of string values of keys to not sanitize.
+
+    :returns: An object with sanitized keys.
+    """
+
+    if ignore_keys is None:
+        ignore_keys = frozenset([])
+
+    deferred_removals = deque()
+
+    no_log_strings = [to_native(s, errors='surrogate_or_strict') for s in no_log_strings]
+    new_value = _sanitize_keys(obj, no_log_strings, ignore_keys, deferred_removals)
+
+    while deferred_removals:
+        old_data, new_data = deferred_removals.popleft()
+
+        if isinstance(new_data, Mapping):
+            for old_key, old_elem in old_data.items():
+                if old_key not in ignore_keys and not old_key.startswith('_ansible'):
+                    # Sanitize the old key. We take advantage of the sanitizing code in
+                    # _remove_values_conditions() rather than recreating it here. Since the
+                    # value should always be a string-like object, it's safe to pass None
+                    # for the deferred_removals parameter.
```

Comment:
A few notes:  This isn't guaranteed to be string-like.

* We could specify that the keys have to be scalar since json dict keys must be scalars.  If we do that, we also need to specify that caveat in the docstring.

Reply:
The stickler in me also thinks we should pass in empty list here, rather than None.  That way, we aren't depending on things which are really internal implementation of the `_remove_values_conditions()` function.

Reply:
Hmmm....... actually, we already created `deferred_removals`....  so there really isn't a cost to passing `deferred_removals` in here, even though we don't expect that anything will be added to it.  So I think we should go ahead and pass it in.

Reply:
Scratch what I just said... if something was added, then it would need to be processed differently than the things that are currently in deferred_removals.  So yeah, back to thinking empty list is the right thing to pass.